custom/custompfm/keeper/keeper.go

```diff
@@ -0,0 +1,311 @@
+// forward keeper and the underlying application.
+type IBCMiddleware struct {
+	router.IBCMiddleware
+
+	app1    porttypes.IBCModule
+	keeper1 *keeper.Keeper
+
+	retriesOnTimeout1 uint8
+	forwardTimeout1   time.Duration
+	refundTimeout1    time.Duration
+	ibcfeekeeper      *ibctransfermiddlewarekeeper.Keeper
+	bank              *custombankkeeper.Keeper
+}
+
+func NewIBCMiddleware(
+	app porttypes.IBCModule,
+	k *keeper.Keeper,
+	retriesOnTimeout uint8,
+	forwardTimeout time.Duration,
+	refundTimeout time.Duration,
+	ibcfeekeeper *ibctransfermiddlewarekeeper.Keeper,
+	bankkeeper *custombankkeeper.Keeper,
+) IBCMiddleware {
+	return IBCMiddleware{
+		IBCMiddleware: router.NewIBCMiddleware(app, k, retriesOnTimeout, forwardTimeout, refundTimeout),
+		ibcfeekeeper:  ibcfeekeeper,
+
+		app1:              app,
+		keeper1:           k,
+		retriesOnTimeout1: retriesOnTimeout,
+		forwardTimeout1:   forwardTimeout,
+		refundTimeout1:    refundTimeout,
+		bank:              bankkeeper,
+	}
+}
+
+func (im IBCMiddleware) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) ibcexported.Acknowledgement {
+	logger := im.keeper1.Logger(ctx)
+
+	var data transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		logger.Debug(fmt.Sprintf("packetForwardMiddleware OnRecvPacket payload is not a FungibleTokenPacketData: %s", err.Error()))
+		return im.IBCMiddleware.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	logger.Debug("packetForwardMiddleware OnRecvPacket",
+		"sequence", packet.Sequence,
+		"src-channel", packet.SourceChannel, "src-port", packet.SourcePort,
+		"dst-channel", packet.DestinationChannel, "dst-port", packet.DestinationPort,
+		"amount", data.Amount, "denom", data.Denom, "memo", data.Memo,
+	)
+
+	d := make(map[string]interface{})
+	err := json.Unmarshal([]byte(data.Memo), &d)
+	if err != nil || d["forward"] == nil {
+		// not a packet that should be forwarded
+		logger.Debug("packetForwardMiddleware OnRecvPacket forward metadata does not exist")
+		return im.app1.OnRecvPacket(ctx, packet, relayer)
+	}
+	m := &types.PacketMetadata{}
+	err = json.Unmarshal([]byte(data.Memo), m)
+	if err != nil {
+		logger.Error("packetForwardMiddleware OnRecvPacket error parsing forward metadata", "error", err)
+		return newErrorAcknowledgement(fmt.Errorf("error parsing forward metadata: %w", err))
+	}
+
+	metadata := m.Forward
+
+	goCtx := ctx.Context()
+	processed := getBoolFromAny(goCtx.Value(types.ProcessedKey{}))
+	nonrefundable := getBoolFromAny(goCtx.Value(types.NonrefundableKey{}))
+	disableDenomComposition := getBoolFromAny(goCtx.Value(types.DisableDenomCompositionKey{}))
+
+	if err := metadata.Validate(); err != nil {
+		logger.Error("packetForwardMiddleware OnRecvPacket forward metadata is invalid", "error", err)
+		return newErrorAcknowledgement(err)
+	}
+
+	// override the receiver so that senders cannot move funds through arbitrary addresses.
```

Comment:
this is not my code. this from original version of IBCMiddleware. so instead of fork all module, i just copy pasted the current function and inject extra code with fee logic.